mdpow/analysis/ensemble.py

```diff
@@ -0,0 +1,499 @@
+from MDAnalysis.lib.log import ProgressBar
+from MDAnalysis.exceptions import FileFormatWarning, NoDataError, MissingDataWarning, SelectionError
+
+from gromacs.utilities import in_dir
+
+from . import NoDataWarning
+
+import logging
+
+logger = logging.getLogger('mdpow._ensemble')
+
+
+class Ensemble(object):
+    """Collection of related MDAnalysis Universe objects
+
+    Stores systems produced by running mdpow-fep organized
+    by solvent, interaction, and lambda.
+
+    Given a mdpow simulation directory will load the MD
+    simulation files with the directory structure as keys.
+
+    :Keywords:
+    *dirname*
+        Molecule Simulation directory. Loads simulation files present in
+        lambda directories into the new instance. With this method for
+        generating an :class:`~mdpow.analysis.ensemble.Ensemble` the
+        lambda directories are explored and
+        :meth:`~mdpow.analysis.ensemble.Ensemble._load_universe_from_dir`
+        searches for .gro, .gro.b2z, .gro.gz, and .tpr files for topology,
+        and .xtc files for trajectory. It will default to using the tpr file
+        available.
+
+    *solvents*
+        Solvents from directory given to the new instance. Default
+        :code:`solvents=('water', 'octanol')`
+
+    *topology_paths*
+        Specifies topologies used in loading simulated systems. Given
+        with a dictionary with keys-value pair for each solvent and
+        its respective topology path.
+
+    *interactions*
+        Interactions from directory given to the instance. Default
+        :code:`interactions=('Coulomb', 'VDW')`
+
+    .. rubric:: Examples
+
+
+    Typical workflow for MDPOW directory::
+
+        ens = Ensemble(dirname='molecule')
+
+    Typical workflow for adding universes individually::
+
+        ens = Ensemble()
+        ens.add_system(topology='md.gro', trajectory='md.xtc')
+
+    Topology paths can be specified when defining the _ensemble
+    by giving the paths to each solvent topology in a dictionary
+    with the topology_paths argument::
+
+        ens = Ensemble(dirname='molecule', topology_paths={'water': water_path,
+                                                           'octanol': octanol_path}
+
+    Interactions can also be specified when initializing the with
+    a list using the interactions argument::
+
+        ens = Ensemble(dirname='molecule', interactions=['Coulomb']
+
+    .. versionadded:: 0.8.0
+    """
+
+    def __init__(self, dirname=None, solvents=('octanol', 'water'),
+                 topology_paths=None, interactions=('Coulomb', 'VDW')):
+        self.top_dict = topology_paths
+        self._num_systems = 0
+        self._ensemble = {}
+        self._keys = []
+        self._solvents = solvents
+        self._interactions = interactions
+
+        if dirname is None:
+            return
+
+        if not os.path.exists(dirname):
+            logger.error(f"Directory {dirname} does not exist")
+            raise FileNotFoundError
```

Comment:
The idea was to raise an exception if the directory structure given when initializing ensemble did not exist, but don't think it's actually needed. If the user gives incorrect settings it will just give an error anyway.